packages/tpetra/core/src/Tpetra_Details_Behavior.cpp

```diff
@@ -84,6 +175,50 @@ namespace { // (anonymous)
     return value;
   }
 
+  bool
+  idempotentlyGetNamedEnvironmentVariableAsBool (const char name[],
+                                                 std::once_flag& once_flag,
+                                                 bool& initialized,
+                                                 const char environmentVariableName[],
+                                                 const bool defaultValue)
+  {
+    using BehaviorDetails::namedVariableMap_;
+    // The extra "initialized" check avoids the cost of synchronizing
+    // on the std::call_once for every call to this function.  We want
+    // it to be cheap to fill the namedVariableMap_, so that users aren't
+    // tempted to try to cache it themselves.
+    if (! initialized) {
+      std::call_once (once_flag, [&] () {
+          setEnvironmentVariableMap (environmentVariableName,
+                                     namedVariableMap_,
+                                     defaultValue);
+          // http://preshing.com/20130922/acquire-and-release-fences/
+          //
+          // "A release fence prevents the memory reordering of any
+          // read or write which precedes it in program order with any
+          // write which follows it in program order."
+          //
+          // The point is to prevent the assignment to 'namedValueMap' from
+          // getting reordered after the assignment to 'initialized' (the
+          // so-called "StoreStore" reordering).  That would be bad in this
+          // case, because then other threads might read 'initialized' as true,
+          // yet would fail to pick up the change to 'namedValueMap'.
+          //
+          // It's harmless if other threads don't see the write to
+          // 'initialized', but did see the write to 'namedValueMap'.  In that
+          // case, they would just attempt and fail to enter the std::call_once,
+          // and return (the correct value of) 'namedValueMap'.
+          std::atomic_thread_fence (std::memory_order_release);
+
+          initialized = true;
+        });
+    }
+    auto thisEnvironmentVariableMap = namedVariableMap_[environmentVariableName];
```

Comment:
This _should_ be a read-only access, because `setEnvironmentVariableMap` should have already added the `environmentVariableName` key to `namedVariableMap_`.  I'm not sure whether `operator[]` and `find` have the same thread-safety guarantees for read-only access, but I think it's fine for now.